#11191 Basic implementation of class Parabola (Geometry module)

ChristinaZografou opened the pull request:
Following a discussion in the mailing list and after having raised an issue about this, we're ready to present you a basic class for Parabola. We are new to contributing and due to academic obligations we weren't able to fully implement every aspect of Parabola, so we focused on the basics. Our implemetation includes the following properties of Parabola: focus, directrix, axis of symmetry, focal length, p parameter, vertex, and the equation of parabola. If the pull request is accepted we are willing to work more on this class and add more functions.

dimitra-kon committed:
Added a new class that implements Parabola

In the geometry module, we added a new class that implements Parabola.
This class is currently supporting only vertical and horizontal parabolas.

Example:

In [1]: from sympy import Parabola, Point, Line

In [2]: p1 = Parabola(Point(0, 0), Line(Point(5, 8), Point(7, 8)))

In [3]: p1.focus

Out [4]: Point2D(0, 0)

In [5]: p1.directrix

Out [6]: Line(Point2D(5, 8), Point2D(7, 8))

Commit:
Added a basic test for class Parabola and made some changes to other classes to include/support Parabola.

In the test file in geometry module, we added a new file "test_parabola.py" that implements basic testing for Parabola.
We also added some testing for Parabola in "test_args.py" in core module.
Apart from that we imcluded Parabola as import in other relative classes.

Example:

$ bin/test test_parabola.py
============================= test process starts ==============================
executable:         C:\Python34\python.exe  (3.4.3-final-0) [CPython]
architecture:       32-bit
cache:              yes
ground types:       python
random seed:        40528811
hash randomization: on (PYTHONHASHSEED=3075141191)

sympy\geometry\tests\test_parabola.py[1] .                                  [OK]

================== tests finished: 1 passed, in 0.66 seconds ===================

jksuom commented on sympy/geometry/tests/test_parabola.py:
```diff
@@ -0,0 +1,59 @@
+                            Point2D, Polygon, Ray, RegularPolygon,
+                            Segment, Triangle, intersection, Parabola)
+from sympy.integrals.integrals import Integral
+from sympy.utilities.pytest import raises, slow
```
It seems that many of the above imports are not necessary for these tests. They should be removed.

Commit:
Removed unnecessary imports in test_parabola.py

By removing unnecessary imports the test is running even faster.

Example:

$ bin/test test_parabola.py
============================= test process starts ==============================
executable:         C:\Python34\python.exe  (3.4.3-final-0) [CPython]
architecture:       32-bit
cache:              yes
ground types:       python
random seed:        26631265
hash randomization: on (PYTHONHASHSEED=2056291977)

sympy\geometry\tests\test_parabola.py[1] .                                  [OK]

================== tests finished: 1 passed, in 0.72 seconds ===================

jksuom commented:
There could be some explanations in the docstring for the less known properties (focal length vs. p parameter, eccentricity) It would also be useful to have a reference (to wikipedia, for example).

sampadsaha5 commented on sympy/geometry/parabola.py:
```diff
@@ -0,0 +1,312 @@
+
+        """
+        x = _symbol(x)
+        y = _symbol(y)
```
Can you please explain me why have you passed 'x' and 'y' as a default value of `x` and `y` respectively while you have also defined them as `_symbol(x)` and `_symbol(y)` respectively?

asmeurer replied:
Is this used in other geometry functions? The input ought to be symbols, not strings.

dimitra-kon replied:
@asmeurer  Yes, the same way is used in ellipse and line. So, do we have to do it in a different way?

sampadsaha5 commented:
Seems a good work.

asmeurer commented on sympy/geometry/parabola.py:
```diff
@@ -0,0 +1,312 @@
+        The eccentricity for every Parabola is 1 by definition.
+
+        """
+        return 1
```
Use `S(1)` so that it returns the sympy type.

asmeurer commented on sympy/geometry/parabola.py:
```diff
@@ -0,0 +1,312 @@
+
+    @property
+    def ambient_dimension(self):
+        return 2
```
Same here. `S(2)`.

asmeurer commented on sympy/geometry/parabola.py:
```diff
@@ -0,0 +1,312 @@
+
+        if len(focus) != 2:
+            raise ValueError('The focus must be a two dimensional'
+                             ' point'.format(cls))
```
I think the format call should not be here.

asmeurer replied:
You can use the `raises` function to test the errors in the test file.

asmeurer commented on sympy/geometry/tests/test_parabola.py:
```diff
@@ -0,0 +1,54 @@
+    assert pa6.eccentricity == 1
+    assert pa7.focus == Point2D(3, 7)
+    assert pa7.focal_length == 1/2
+    assert pa7.p_parameter == -1/2
```
Use sympy rationals in the test file.

sampadsaha5 commented on sympy/geometry/tests/test_parabola.py:
```diff
@@ -0,0 +1,54 @@
+    assert pa8.focal_length == pa9.focal_length
+    assert pa8.p_parameter == pa9.p_parameter
+    assert pa8.vertex == pa9.vertex
+    assert pa8.equation() == pa9.equation()
```
Please add tests for the raised errors.

sampadsaha5 commented on sympy/geometry/parabola.py:
```diff
@@ -0,0 +1,312 @@
+from sympy.geometry.line import Line
+from sympy.geometry.util import _symbol
+
+import random
```
You can remove this import since `random` is not used in this module.

sampadsaha5 commented on sympy/geometry/tests/test_parabola.py:
```diff
@@ -0,0 +1,54 @@
+from __future__ import division
+
+from sympy import oo
+from sympy.geometry import Line, Point, Point2D, Parabola
```
`Line, Point, Point2D, Parabola` can be imported directly from sympy.

dimitra-kon replied:
@asmeurer  @sampadsaha5 Thank you for the feedback. We are working to correct them!

dimitra-kon committed:
Made some significant changes in parabola.py and test_parabola.py

In parabola.py :   Added some explanations in the docstrings for the less known properties, used S() to return the sympy type, removed a format call, removed an unused import of random, added a condition to check if focus is a point of directrix and changed the input type of equation() function.

In test_parabola.py: Added tests for raised errors, used sympy rationals and Line, Point, Point2D, Parabola are imported directly from sympy.

Commit:
Changed docstring in equation() in order to fix error causing Travis checks to fail

Example:

In [1]: from sympy import Parabola, Point, Line, Symbol
In [2]: p1 = Parabola(Point(0, 0), Line(Point(5, 8), Point(7, 8)))
In [3]: p1.equation()
Out [4]: -x**2 - 16*y + 64
In[5]: p1.equation(Symbol("f"))
Out[6]: -f**2 - 16*y + 64
In[7]: p1.equation(y = Symbol("z"))
Out[8]:  -x**2 - 16*z + 64

moorepants commented on sympy/geometry/parabola.py:
```diff
@@ -0,0 +1,359 @@
+
+        """
+        if x is None:
+            x = Symbol("x")
```
This would be better if it was sympify('x').

Also, these variables should be attributes to the class. See issue #11043. Users may need access to these and there isn't a simple way to get them. I would suggest initialializing them in the `__init__` method as attributes. Then change those attributes here if different ones are passed in by the user. Or get rid of these options all together. The user can do this:

```python
p = Parabola(...)
p.ordinate_var = symbols('y')
p.abscissa_var = symbols('x')
p.equation()
```

asmeurer replied:
Why sympify('x')? That should give the same thing.

asmeurer replied:
The equation method here should indeed match the other stuff in the geometry module. Were the other equation methods added before or after sympy 1.0?

moorepants replied:
> Why sympify('x')? That should give the same thing.

I thought sympify handles a more general set of inputs than Symbol. But I suppose you only want symbol-like objects here. The `_symbol()` function was here earlier (and in the other geometry classes) and that makes the symbols have the real assumption, which complicates the issues in #11043 a bit  more.

> The equation method here should indeed match the other stuff in the geometry module.

The design of this is not that great in my opinion. Especially wrt to the issue in #11043.

> Were the other equation methods added before or after sympy 1.0?

Before I believe.

moorepants replied:
Oh, darn. My above suggestion would break immutability, at least if you could set the `abscissa_var` and `ordinate_var`.

asmeurer replied:
I didn't know that about `_symbol`. If the other geometry functions use that, then this should as well. 

Regarding the design, let's just match the rest of the geometry module here, and fix the design in a future issue.

moorepants replied:
> Regarding the design, let's just match the rest of the geometry module here, and fix the design in a future issue. 

Sounds fine.

dimitra-kon replied:
@asmeurer @moorepants We have to confess that we are a little bit confused. Finally, what exactly you want us to do? We have to change something after our latest commit?

moorepants replied:
You should copy the form of the `equation()` method in the other classes in the geometry module. I'm pretty sure they use the `_symbol()` function here which is in the geometry util module.

dimitra-kon replied:
Ok, that's what we have done first and we were told to change it. We are going to use _symbol() again!

asmeurer replied:
Yes, sorry. I didn't realize the other methods in the geometry module did this.

dimitra-kon replied:
It's ok! Do you think that there is something else that we should change with our next commit?

asmeurer replied:
Everything else looks good to me. I was going to wait until Monday to see if anyone else wants to review and merge if there are no more comments.

Commit:
Changes in the variables of the equation():

        """The equation of the parabola.

        Parameters
        ==========
        x : str, optional
            Label for the x-axis. Default value is 'x'.
        y : str, optional
            Label for the y-axis. Default value is 'y'.

        Returns
        =======
        equation : sympy expression

        Examples
        ========

        >>> from sympy import Parabola, Point, Line
        >>> p1 = Parabola(Point(0, 0), Line(Point(5, 8), Point(7, 8)))
        >>> p1.equation()
        -x**2 - 16*y + 64
        >>> p1.equation('f')
        -f**2 - 16*y + 64
        >>> p1.equation(y='z')
        -x**2 - 16*z + 64

        """

smichr commented on sympy/geometry/parabola.py:
```diff
@@ -0,0 +1,352 @@
+class Parabola(GeometrySet):
+    """A parabolic GeometryEntity.
+
+    A parabola is declared with a point, that is called 'focus', and
```
Other instantiations could be implemented (now or later): vertex and point or three points and direction or no arguments to create the canonical vertical parabola (`y=x**2`) -- like the canonical triangle created as `Triangle(sss=(1,1,1))` -- and then create a specific case with rotation or scaling.

ChristinaZografou replied:
Thank you for your feedback. Yes we had thought about that but we didn't have the time to dive deep in it. We focused on some basic stuff and implementing them right but we definitely intend to add more functionalities or options to the already existing ones once we have more time in our hands.

smichr commented:
Don't know if anyone will see this, but #7803 might have something of interest. I haven't looked, but it was a PR for a Parabola class, too.
